keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,242 @@
+
+## Motivation
+
+Current scheduler predicates for scheduling of pods with volumes is based on `node.status.capacity` and `node.status.allocatable`. It works well for hardcoded predicates for volume limits on AWS (`MaxEBSVolumeCount`), GCE(`MaxGCEPDVolumeCount`), Azure (`MaxAzureDiskVolumeCount`) and OpenStack (`MaxCinderVolumeCount`).
+
+It is problematic for CSI (`MaxCSIVolumeCountPred`) outlined in [#730](https://github.com/kubernetes/enhancements/pull/730)
+
+- `ResourceName` is limited to 63 characters. We must prefix `ResourceName` with unique string (such as `attachable-volumes-csi-<driver name>`) so it cannot collide with existing resources like `cpu` or `memory`. But `<driver name>` itself is up to 63 character long, so we ended up with using SHA-sums of driver name to keep the `ResourceName` unique, which is not user readable.
+- CSI driver cannot share its limits with in-tree volume plugin e.g. when running pods with AWS EBS in-tree volumes and `ebs.csi.aws.com` CSI driver on the same node.
+- `node.status` size increases with each installed CSI driver. Node objects are big enough already.
+
+### Goals
+
+- User can run use PVs both with in-tree volume plugins and CSI and they will share their limits. There is only one scheduler predicate that handles both kind of volumes.
+
+- Existing predicates for in-tree volumes `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` are removed (with deprecation period).
+  - When both deprecated in-tree predicate and CSI predicate are enabled, only one of them does useful work and the other is NOOP to save CPU.
+
+- Scheduler does not increase its CPU consumption. Any regression must be approved by sig-scheduling.
+
+### Non-Goals
+
+- Heterogenous clusters, i.e. clusters where access to storage is limited only to some nodes. Existing `PV.spec.nodeAffinity` handling, not modified by this KEP, will filter out nodes that don't have access to the storage, so predicates changed in this KEP don't need to worry about storage topology and can be simpler.
+
+## Proposal
+
+* Track volume limits for both in-tree volume plugins and CSI drivers in `CSINode` objects instead of `Node`.
+  * To get rid of prefix + SHA for `ResourceName` of CSI volumes.
+  * So in-tree volume plugin can share limits with CSI driver that uses the same storage backend.
+
+* Kubelet will create `CSINode` instance during initial node registration together with `Node` object.
+  * Limits of each in-tree volume plugin will be added to `CSINode.status.allocatable`.
+    * Limit for in-tree volumes will be added by kubelet during CSINode creation. Name of corresponding CSI driver will be used as key in `CSINode.status.allocatable` and it will be discovered using [CSI translation library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib). If the library does not support migration of an in-tree volume plugin, the volume plugin has no limit.
```

Comment:
Does this mean the translation library will need the cloud provider logic that currently determines the attach limit? Such as [this](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gcepd/gce_pd.go#L126)?

Reply:
No. All in-tree volume plugins already have `GetVolumeLimits` call. When the plugin is still supported in-tree, we get the limit from there. If there is only CSI, we get the limit from the CSI driver, meaning that the initial `CSINode` may have no limits, because no drivers are registered yet.

Reply:
It might be good to have a section that clearly spells out what happens when:
* Migration is off
* Migration is on but driver not installed
* Migration is on and driver is installed
* In-tree is removed, and migration is always on, and driver is/is not installed

Reply:
Added to Design Details.